client/modules/IDE/pages/IDEView.jsx

```diff
@@ -66,296 +70,212 @@ function WarnIfUnsavedChanges() {
+  const isUserOwner = useSelector(getIsUserOwner);
+  const dispatch = useDispatch();
+  const { t } = useTranslation();
 
-    this.state = {
-      consoleSize: props.ide.consoleIsExpanded ? 150 : 29,
-      sidebarSize: props.ide.sidebarIsExpanded ? 160 : 20
-    };
-  }
+  const [consoleSize, setConsoleSize] = useState(150);
+  const [sidebarSize, setSidebarSize] = useState(160);
+  const [isOverlayVisible, setIsOverlayVisible] = useState(true);
 
-  componentDidMount() {
-    // If page doesn't reload after Sign In then we need
-    // to force cleared state to be cleared
-    this.props.clearPersistedState();
+  const cmRef = useRef({});
 
-    this.props.stopSketch();
-    if (this.props.params.project_id) {
-      const { project_id: id, username } = this.props.params;
-      if (id !== this.props.project.id) {
-        this.props.getProject(id, username);
-      }
-    }
+  const autosaveIntervalRef = useRef(null);
 
-    // window.onbeforeunload = this.handleUnsavedChanges;
-    window.addEventListener('beforeunload', this.handleBeforeUnload);
+  const location = useLocation();
 
-    this.autosaveInterval = null;
-  }
+  const prevFileNameRef = useRef(selectedFile.name);
+  const locationRef = useRef(location.pathname);
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.location !== this.props.location) {
-      this.props.setPreviousPath(this.props.location.pathname);
-    }
-    if (this.props.ide.sidebarIsExpanded !== nextProps.ide.sidebarIsExpanded) {
-      this.setState({
-        sidebarSize: nextProps.ide.sidebarIsExpanded ? 160 : 20
-      });
-    }
-  }
+  const syncFileContent = () => {
+    const file = cmRef.current.getContent();
+    dispatch(updateFileContent(file.id, file.content));
+  };
+
+  useEffect(() => {
+    dispatch(clearPersistedState());
 
-  componentWillUpdate(nextProps) {
-    if (nextProps.params.project_id && !this.props.params.project_id) {
-      if (nextProps.params.project_id !== nextProps.project.id) {
-        this.props.getProject(nextProps.params.project_id);
+    dispatch(stopSketch());
+    if (props.params.project_id) {
+      const { project_id: id, username } = props.params;
+      if (id !== project.id) {
+        dispatch(getProject(id, username));
       }
     }
-  }
+  }, []);
+
+  // for setting previous location
+  useEffect(() => {
+    dispatch(setPreviousPath(locationRef.current.pathname));
+  }, [location.pathname]);
 
-  componentDidUpdate(prevProps) {
-    if (this.props.isUserOwner && this.props.project.id) {
+  // For autosave
+  useEffect(() => {
+    const handleAutosave = () => {
       if (
-        this.props.preferences.autosave &&
-        this.props.ide.unsavedChanges &&
-        !this.props.ide.justOpenedProject
+        isUserOwner &&
+        project.id &&
+        preferences.autosave &&
+        ide.unsavedChanges &&
+        !ide.justOpenedProject &&
+        selectedFile.name === prevFileNameRef.current
```

Comment:
As we discussed, we can do the conditional checks before we start the timer.  So we don't need any `if` here.
```suggestion
```